### PR TITLE
Test Framework Setup + Convergence Score Definition

### PR DESCRIPTION
**Pull request by weihangzheng**
## Enhancements to Simulation Testing Framework

---

### Introduction
This pull request is a work-in-progress that introduces preliminary enhancements to our existing Simulation Testing Framework, focusing on the `SimulationTester` class. 

### Changes Made
1. **Enhanced Initialization**: Updated the `__init__` method of the `SimulationTester` class to accept an `included_data` parameter. This parameter is a list of strings indicating which data components (like `observations`, `rewards`, etc.) to include in the simulation results. This change increases customization capabilities for simulation runs.

2. **Simulation Execution Method**: Added `run_simulation` method to execute the simulation and store results. 

3. **Convergence Score Functionality**: Introduced `calculate_convergence_score` method, based on a cost function modified from Diane and Prof. Pavel's paper. This method calculates the convergence score from the average reward across all agents in the last iteration of the algorithm. **An ideal / globally optimal NE will have a score of zero. Lower the score the closer it is to globally optimal NE.**

4. **Results Display**: Implemented `print_results` method to print the simulation results stored from `run_simulation`. 
5. **Changed Global Arrangement Cost (GAC)**: This was a part of the rewards function and it needed to be changed as we had previously relied on average Euclidean distance from the target. This was misleading. Say two agents were on the opposite side of the target by x units, the GAC would be 2x previously. In reality, it should be x - x = 0. This helps to ensure that the globally optimal NE always has a convergence score of 0.

### Example Usage
Included at the end of the `SimulationTester` class file is an example usage scenario demonstrating how to initialize the tester, run a simulation, and print the results.

### Files Changed
- `baseline.py`: Modified to integrate with the new `SimulationTester` class.
- `nepiada.py`: Updated to support changes in simulation data components handling.
- `test.py`: Refactored to align with the enhanced data component selection in `SimulationTester`.

### Testing
Initial tests were conducted to ensure basic functionality. However, as this is a work-in-progress, further testing and refinement are required.

Sanity checks conducted:

- Convergence score is indeed the average of rewards of the last iteration step.
- Plotted a global arrangement vector to visually confirm that we are computing the correct rewards

---

### CODE REVIEW

**Comment by pandyah5** [comment]
Hi @weihangzheng , I really like this test setup, I'll polish and test this a bit to address #39

**Comment by pandyah5** [comment]
@weihangzheng When you work on "Write test cases for the test framework", can you also look into converting the print output to a CSV file? Doing so can help us produce good graphics in Excel for the meetings 4 and 5.

File: src/nepiada/env/nepiada.py [hunk 1]
```diff
@@ -370,19 +373,20 @@ def get_rewards(self):
...
-        global_arrangement_reward = 0
+        global_arrangement_vector = np.array([0.0, 0.0])
 
         for agent_name in self.agents:
             agent = self.world.agents[agent_name]
-            distance = np.sqrt(
-                (agent.p_pos[0] - target_x) ** 2 + (agent.p_pos[1] - target_y) ** 2
-            )
-            global_arrangement_reward += distance
+            global_arrangement_vector += np.array([(agent.p_pos[0] - target_x), (target_y - agent.p_pos[1])])
```
Comment (pandyah5) [inline src/nepiada/env/nepiada.py:383]: This is a very important change so I would like a second opinion here.

**Review by Thanoshan — COMMENTED** [review]
Looks good overall!

File: src/nepiada/env/nepiada.py (hunk 1 above)
Comment (pandyah5) [reply src/nepiada/env/nepiada.py:383]: Yup, I'll modify them in my next commit

File: src/nepiada/test.py
[the remark this answers is not in the document]
Comment (pandyah5) [reply src/nepiada/test.py:1]: Addressed in [727d9f1](https://github.com/BoundlessDevelopment/Capstone-Project/pull/43/commits/727d9f1b47c9d98ff494467d11971e87aebdcdf6)

File: src/nepiada/test.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Import the main function from the simulation script
+from baseline import main
```
[the remark this answers is not in the document]
Comment (pandyah5) [reply src/nepiada/test.py]: Addressed in [727d9f1](https://github.com/BoundlessDevelopment/Capstone-Project/pull/43/commits/727d9f1b47c9d98ff494467d11971e87aebdcdf6)

File: src/nepiada/env/nepiada.py (hunk 1 above)
Comment (pandyah5) [reply src/nepiada/env/nepiada.py:383]: Addressed in [727d9f1](https://github.com/BoundlessDevelopment/Capstone-Project/pull/43/commits/727d9f1b47c9d98ff494467d11971e87aebdcdf6)

**Review by Thanoshan — APPROVED** [review]
LGTM. Thanks Hetav!

**Pull request merged**